sponnet/pipeline.libsonnet

```diff
@@ -2,52 +2,118 @@
     triggers: [],
-    withApplication(application):: self + {application: application},
-    withName(name):: self + {name: name},
-    withStages(stages):: self + if std.type(stages) == "array" then {stages: stages} else {stages: [stages]},
-    withTriggers(triggers):: self + if std.type(triggers) == "array" then {triggers: triggers} else {triggers: [triggers]},
+    withApplication(application):: self + { application: application },
+    withExpectedArtifacts(expectedArtifacts):: self + if std.type(expectedArtifacts) == 'array' then { expectedArtifacts: expectedArtifacts } else { expectedArtifacts: [expectedArtifacts] },
+    withName(name):: self + { name: name },
+    withNotifications(notifications):: self + if std.type(notifications) == 'array' then { notifications: notifications } else { notifications: [notifications] },
+    withStages(stages):: self + if std.type(stages) == 'array' then { stages: stages } else { stages: [stages] },
+    withTriggers(triggers):: self + if std.type(triggers) == 'array' then { triggers: triggers } else { triggers: [triggers] },
   },
 
   moniker(app, cluster):: {
     app: app,
     cluster: cluster,
-    withStack(stack):: self + {stack: stack},
-    withDetail(detail):: self + {detail: detail},
+    withStack(stack):: self + { stack: stack },
+    withDetail(detail):: self + { detail: detail },
   },
 
   // artifacts
 
   artifact(type):: {
     type: type,
-    withName(name):: self + {name: name},
-    withVersion(version):: self + {version: version},
-    withReference(reference):: self + {reference: reference},
-    withLocation(location):: self + {location: location},
+    withName(name):: self + { name: name },
+    withVersion(version):: self + { version: version },
+    withReference(reference):: self + { reference: reference },
+    withLocation(location):: self + { location: location },
   },
 
   local artifact = self.artifact,
   artifacts:: {
-    dockerImage():: artifact("docker/image"),
-    githubFile():: artifact("github/file"),
-    gcsObject():: artifact("gcs/object"),
+    dockerImage():: artifact('docker/image'),
+    githubFile():: artifact('github/file'),
+    gitlabFile():: artifact('gitlab/file'),
+    gcsObject():: artifact('gcs/object'),
   },
 
+  // expected artifacts
+  // TODO: This section may need splitting out by artifact type due to differing field requirements.
+
   expectedArtifact(id):: {
     id: id,
-    withMatchArtifact(matchArtifact):: self + {matchArtifact: matchArtifact},
-    withDefaultArtifact(defaultArtifact):: self + {defaultArtifact: defaultArtifact},
-    withUsePriorArtifact(usePriorArtifact):: self + {usePriorArtifact: usePriorArtifact},
-    withUseDefaultArtifact(useDefaultArtifact):: self + {useDefaultArtifact: useDefaultArtifact},
+    // Defaults to kind: custom
+    defaultArtifact: { kind: 'custom' },
+    withMatchArtifact(matchArtifact):: self + {
+      matchArtifact+: {
+        kind: std.splitLimit(matchArtifact.type, '/', 1)[0],
```

Comment:
Kind is just for the UI, and doesn't (as far as I know) always follow this pattern. I'd drop this.

Reply:
Dropped.

Reply:
https://github.com/spinnaker/spinnaker/issues/3491